### PR TITLE
createSupplier: Hoist call to getConstructor.

### LINKED ISSUE #820: SupplierUtil.createSupplier returns a slow supplier
**Issue**

https://github.com/jgrapht/jgrapht/blob/master/jgrapht-core/src/main/java/org/jgrapht/util/SupplierUtil.java#L58

Calls getDeclaredConstructor every time.

**Expected behaviour**

Hoist the call to getDeclaredConstructor and close over it.
```
    @Nonnull
    public static <T> Supplier<T> createSupplier(Class<? extends T> edgeClass) {
        try {
            Constructor<? extends E> constructor = edgeClass.getDeclaredConstructor();
            return () -> {
                try {
                    return constructor.newInstance();
                } catch (ReflectiveOperationException e) {
                    throw new RuntimeException(e);
                }
            };
        } catch (ReflectiveOperationException | SecurityException e) {
            throw new RuntimeException(e);
        }
    }
```

**Conclusions**

LOL another performance issue from those random idiots who are doing WTF? :D :D

### PR DESCRIPTION
**Pull request by shevek**
Fixes #820 
----

- [ ] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [ ] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [ ] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [ ] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [ ] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [ ] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [ ] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

**Comment by shevek** [comment]
This fails a test because someone calls createSupplier(Integer.class) somewhere, which is a preexisting bug and not the fault of this PR.

**Comment by shevek** [comment]
This is more complicated than the code pasted into #820  (which is what we use internally) because I was careful to honour the Serializable contract as well for you, which we don't use.

**Comment by jsichi** [comment]
Regarding the remaining test failures, the PR needs to be updated to avoid them.  I get what you are saying about pre-existing conditions, but since we never actually deprecated the edge-class-based constructors for all the classes such as SimpleGraph, the PR as is would be an unacceptable breaking change.

The failures occur because a lazy operation has been converted into an eager operation.  So maybe the simplest fix is to handle the eager error by returning a supplier which will fail if it's ever invoked.

File: jgrapht-core/src/main/java/org/jgrapht/util/SupplierUtil.java [hunk 1]
```diff
@@ -48,6 +50,46 @@
...
+    private static class ConstructorSupplier<T> implements Supplier<T>, Serializable {
+
+        private static class SerializedForm implements Serializable {
+            private final Class<?> type;
+
+            public SerializedForm(Class<?> type) {
+                this.type = type;
+            }
+
+            /* pp */ Object readResolve() throws ObjectStreamException {
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/util/SupplierUtil.java]: Does "pp" mean package-private?  No need for the comment.

**Comment by d-michail** [comment]
I did what John suggested in order to fix the tests with the integer class.

In order to speed things up, I copied your code along with the extra fixes into a new PR #868.

**Pull request closed without merging**